### PR TITLE
Port account-optimization examples to GitHub.

### PR DESCRIPTION
**Pull request by wfansh**

### CODE REVIEW

**Comment by wfansh** [comment]
@jradcliff @AnashOommen 

Do you mind helping review this PR?

**Comment by wfansh** [comment]
@jradcliff @AnashOommen 

Do you mind helping review this PR?

**Comment by wfansh** [comment]
These examples combine the usages of `optiScore` and `recommendation` together.

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/README.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    ```
+    CID,AccountName,OldOptiScore,NewOptiScore
+    111111,Overstock - PLA,0.8225589,0.8282797
+    ```
+
+    *   **optimization_111111/recommendations.csv**:
+
+    ```
+    ID,Type,Details,CampaignId,OldOptiScore,NewOptiScore
+    444444,CAMPAIGN_BUDGET,Increase budget(in micros) from 5750000000 to 8400000000,0,0.0,0.0
```
Comment (jradcliff) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/README.md:219]: Is the `0` before `,0.0,0.0` the campaign ID?

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/AcquireOptimizations.java [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+    "BaseVideoViews",
+    "PotentialVideoViews"
+  };
+
+  private static class AcquireOptimizationsParams extends CodeSampleParams {
+
+    @Parameter(
+        names = ArgumentNames.CUSTOMER_IDS,
+        description =
+            "Optionally specify a comma-separated list of customer IDs to retrieve recommendations"
```
Comment (jradcliff) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/AcquireOptimizations.java:88]: It'd be helpful if these descriptions indicated they need to either specify one or the other.

So perhaps:

`"Specify a comma-separated list of customer IDs to retrieve recommendations from. You must provide at least one of " + ArgumentNames.CUSTOMER_IDS + " or " + ArgumentNames.LOGIN_CUSTOMER_ID + "."`

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/AcquireOptimizations.java [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+  Customer getCustomerInfo(GoogleAdsServiceClient googleAdsServiceClient, long customerId) {
+    String query =
+        "SELECT customer.id, customer.descriptive_name, customer.optimization_score FROM customer";
+
+    SearchPagedResponse response = googleAdsServiceClient.search(String.valueOf(customerId), query);
+    for (GoogleAdsRow googleAdsRow : response.iterateAll()) {
+      return googleAdsRow.getCustomer();
+    }
+
+    throw new RuntimeException("Customer not found with ID " + customerId);
```
Comment (jradcliff) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/AcquireOptimizations.java:300]: Use `IllegalArgumentException` instead.

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/ApplyRecommendations.java [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+    if (customerIds == null) {
+      applyRecommendationsToAllCustomers(googleAdsClient, reportDirectory);
+    } else {
+      int numRecommendations = 0;
+      for (long customerId : customerIds) {
+        numRecommendations +=
+            applyRecommendationsToCustomer(googleAdsClient, customerId, reportDirectory);
+      }
+      System.out.printf(
+          "Applied %d recommendations totally to %d customers.%n",
```
[the remark this answers is not in the document]
Comment (wfansh) [reply google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/ApplyRecommendations.java:182]: Done.

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/ApplyRecommendations.java [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+            applyRecommendationsToCustomer(googleAdsClient, customerId, reportDirectory);
+        numCustomers++;
+      } else {
+        numSkips++;
+        System.out.println(file + " is skipped.");
+      }
+    }
+
+    System.out.printf(
+        "Skipped %d files totally in %s.%nApplied %d recommendations totally to %d customers.%n",
```
[the remark this answers is not in the document]
Comment (wfansh) [reply google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/ApplyRecommendations.java:213]: Done

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/ApplyRecommendations.java [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+   * AcquireOptimizations.java.
+   *
+   * @param csvPath the path to 'recommendations.csv' of the recommendation reports.
+   * @param customerId the client customer ID.
+   * @return a map between a recommendation ID and the related {@link AggregatedRecommendation}.
+   * @throws IOException if a failure occurs when trying to read CSV files.
+   */
+  private Map<String, AggregatedRecommendation> loadRetrievedRecommendations(
+      Path csvPath, long customerId) throws IOException {
+    Map<String, AggregatedRecommendation> aggregatedRecommendatios = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (wfansh) [reply google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/ApplyRecommendations.java:309]: Good catch. Done and thanks!

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/README.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+*   Generate the ***recommendation reports*** in the CSV format.
+
+*   Parse the *recommendation reports* after manual review & revision.
+
+*   Apply the recommendations parsed from the reports of the specified Google
+    Ads accounts, and generate the ***optimization reports*** in the CSV format
+    with the apply results and the latest optimization scores.
+
+Before executing the code examples, make sure that you have an `ads.properties`
+file configured properly with your credentials. See
```
[the remark this answers is not in the document]
Comment (wfansh) [reply google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/README.md:20]: Done.

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/README.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    |-- recommendations.csv
+    ```
+
+    The content of the CSV files are as:
+
+    *   **optimization_111111/optiScore.csv**:
+
+    ```
+    CID,AccountName,OldOptiScore,NewOptiScore
+    111111,Overstock - PLA,0.8225589,0.8282797
```
[the remark this answers is not in the document]
Comment (wfansh) [reply google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/README.md:212]: Sounds great.

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/README.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+            ]
+            ```
+
+        *   **SITELINK_EXTENSION**
+
+            ```
+            recommended_extensions: [link_text: "About"
+            line1: ""
+            line2: ""
+            final_urls: "https://about.linkedin.com/?trk=public_profile_v3_mobile_footer-about"
```
[the remark this answers is not in the document]
Comment (wfansh) [reply google-ads-examples/src/main/java/com/google/ads/googleads/examples/optimization/README.md:148]: Changed to "http://example.com/about" and "http://example.com/join" instead.

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/recommendations/optimizations/ApplyRecommendations.java [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @return the customer's optimization score.
+   */
+  private double getCustomerOptiScore(GoogleAdsClient googleAdsClient, long customerId) {
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      return new AcquireOptimizations()
+          .getCustomerInfo(googleAdsServiceClient, customerId)
```
Comment (fiboknacky) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/recommendations/optimizations/ApplyRecommendations.java:415]: This seems not needed to be an instance method. Do we have any location that we can place this method and share it between two files?

**Comment by jradcliff** [comment]
Will revisit this as needed, but closing for now.

**Pull request closed without merging**